### PR TITLE
fuzz: add DHCP server fuzzer

### PR DESCRIPTION
**Pull request by titanous**

### CODE REVIEW

File: src/fuzz/fuzz-dhcp-server.c [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        static const uint8_t chaddr[] = {3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3};
+        uint8_t *client_id;
+        DHCPLease *lease;
+        int r, pool_offset;
+
+        if (size < sizeof(DHCPMessage))
+                return 0;
+
+        r = sd_dhcp_server_new(&server, 1);
+        assert(r >= 0);
```
Comment (yuwata) [inline src/fuzz/fuzz-dhcp-server.c]: I do not know much about fuzz tests, but I think we should use assert_se(). I mean
`assert_se(sd_dhcp_server_new(...) >= 0);`

File: src/fuzz/fuzz-dhcp-server.c [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        /* set fd to a fake value to bypass assert() in dhcp_server_send_udp */
+        server->fd = 128;
+        r = sd_dhcp_server_configure_pool(server, &address, 24, 0, 0);
+        assert(r >= 0);
```
Comment (yuwata) [inline src/fuzz/fuzz-dhcp-server.c]: Same here.

File: src/fuzz/fuzz-dhcp-server.c [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+        /* add a lease to the pool to expose additional code paths */
+        client_id = malloc(2);
```
Comment (yuwata) [inline src/fuzz/fuzz-dhcp-server.c]: Missing allocation failure check.

File: src/fuzz/fuzz-dhcp-server.c [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        client_id[0] = 2;
+        client_id[1] = 2;
+        lease = new0(DHCPLease, 1);
```
Comment (yuwata) [inline src/fuzz/fuzz-dhcp-server.c]: Same here.

File: src/fuzz/fuzz-dhcp-server.c [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        lease->client_id.length = 2;
+        lease->client_id.data = client_id;
+        lease->address = htobe32(UINT32_C(10) << 24 | UINT32_C(2));
+        lease->gateway = htobe32(UINT32_C(10) << 24 | UINT32_C(1));
+        lease->expiration = UINT64_MAX;
+        memcpy(lease->chaddr, chaddr, 16);
+        pool_offset = get_pool_offset(server, lease->address);
+        server->bound_leases[pool_offset] = lease;
+        hashmap_put(server->leases_by_client_id, &lease->client_id, lease);
```
Comment (yuwata) [inline src/fuzz/fuzz-dhcp-server.c]: Check the return value with assert_se().

**Comment by titanous** [comment]
Comments addressed.

File: src/fuzz/fuzz-dhcp-server.c (hunk 4 above)
Comment (titanous) [reply src/fuzz/fuzz-dhcp-server.c]: Done.

File: src/fuzz/fuzz-dhcp-server.c [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        if (size < sizeof(DHCPMessage))
+                return 0;
+
+        assert_se(sd_dhcp_server_new(&server, 1) >= 0);
+        /* set fd to a fake value to bypass assert() in dhcp_server_send_udp */
+        server->fd = 128;
+        assert_se(sd_dhcp_server_configure_pool(server, &address, 24, 0, 0) >= 0);
+
+        /* add a lease to the pool to expose additional code paths */
+        client_id = malloc(2);
```
Comment (yuwata) [inline src/fuzz/fuzz-dhcp-server.c]: Is it not necessary to check this?

Comment (titanous) [reply src/fuzz/fuzz-dhcp-server.c]: Oops, fixed.

Comment (yuwata) [reply src/fuzz/fuzz-dhcp-server.c]: Thanks.

File: src/fuzz/fuzz-dhcp-server.c [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+ssize_t sendmsg(int sockfd, const struct msghdr *msg, int flags) {
+        return 0;
+}
+
+int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
+        _cleanup_(sd_dhcp_server_unrefp) sd_dhcp_server *server = NULL;
+        struct in_addr address = {.s_addr = htobe32(UINT32_C(10) << 24 | UINT32_C(1))};
+        static const uint8_t chaddr[] = {3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3};
```
Comment (keszybz) [inline src/fuzz/fuzz-dhcp-server.c:35]: Hmm, why is this `static`?

File: src/fuzz/fuzz-dhcp-server.c [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+        uint8_t *client_id;
+        DHCPLease *lease;
+        int pool_offset;
+
+        if (size < sizeof(DHCPMessage))
+                return 0;
+
+        assert_se(sd_dhcp_server_new(&server, 1) >= 0);
+        /* set fd to a fake value to bypass assert() in dhcp_server_send_udp */
+        server->fd = 128;
```
Comment (keszybz) [inline src/fuzz/fuzz-dhcp-server.c]: Please don't do that. Sooner or later something will try to close this fd, leading to trouble.
```C
   server->fd = open("/dev/null", O_RDWR|O_CLOEXEC|O_NOCTTY);
   if (server->fd < 0)
               // error handling
```
and don't set it to -1 below.

File: src/fuzz/fuzz-dhcp-server.c (hunk 7 above)
Comment (titanous) [reply src/fuzz/fuzz-dhcp-server.c:35]: It never gets changed and is the same across functions, and my understanding is that `static` will avoid putting it on the stack for every call.

File: src/fuzz/fuzz-dhcp-server.c (hunk 8 above)
Comment (titanous) [reply src/fuzz/fuzz-dhcp-server.c]: Done.

**Pull request merged**